pkg/keycloak/keycloak_handler.go

```diff
@@ -0,0 +1,50 @@
+package keycloak
```

Comment:
1. There should not be any Keycloak specific code. We should not depend on the identity provider implementation unless we really have to.
Right now RHD supports OAuth2 and OpenID Connect specs. And it happens to run Keycloak. But we don't care.

2. Why would we want to have such endpoint to provide information about the Identity Provider in backend at all? Backend doesn't use this information. It's purely for UI. We could keep it in backend if we think we may need to use different clients which will consume such information (like different UIs/front-ends). So, was it your reasoning? Or something else?

Reply:
Afaik we will need to have Keycloak-specific code in the project. While the JWT validation itself is generic, retrieving the public key is not imo.

Reply:
Also, we need to supply the config to the JS Keycloak client. This can be done by:

1. burn the config static into the JS.
2. have it served as a static file from the backend.
3. have it generated from the backend config and served to the UI on an endpoint.

(1) is bad, it would mean to have configurable values compiled in or at least on two different locations (and not configurable using the config system, environment etc).

(2) is almost as bad as (1), it is just externalized from the JS, everything else from (1) applies as well.

(3) would be the way to have those config values unified with the rest of the config and managed and exposed in a sane way. BUT it requires to serve a Keycloak JSON from the backend.

I'd definitely say that a unified, managable config is more important than a generic backend. Especially because the UI *is* Keycloak-specific and an atomic part of the whole application.

Reply:
If we want to go generic, we should dump that Keycloak client and use the standard OAuth2 flow (which was what we started with). But I admit the Keycloak JS client makes things much easier :-)

Reply:
i think there are instances where static data that is purely used only for the ui are served via the back end. this can be seen in the static content hosting pattern for instance where style sheets, html pages, and even static drop down menus are populated via an endpoint. i think in our case, this json is very small and it can either sit in the ui or go in the back end. personally, i prefer to have it served..

Reply:
> While the JWT validation itself is generic, retrieving the public key is not imo.

Retrieving the public keys is not Keycloak specific. It's JWK spec (see https://tools.ietf.org/html/rfc7517).
We have nothing KC specific in auth service and it works fine with RHD which runs Keycloak. But if RHD tomorrow switches to another implementation which still supports Oauth2, OpenID Connect, JWT, JWK, etc. specs then we should be good.

> Especially because the UI is Keycloak-specific and an atomic part of the whole application.

No. It's not. As I already mentioned multiple times we are going to use Keycloak JS Adapter not because it's KC specific. But because it actually works with services that implement Oauth2, OpenID Connect, JWT, JWK, etc. just fine. In addition to that the adapter has KC specific functionality too but we don't want/need to use those functionality. The generic part of KC Adapter is enough for us.
Many moons ago KC JS Adapter worked only with KC. But then we (OSIO-che team) modified it (contributed to upstream) so it can work with any auth service which implements the specs I mentioned above. For example we now use that JS adapter in che.openshift.io which talks to auth.openshift.io (our own go service which implements these specs. Not Keycloak!). We could use something else if we really wanted but why not using KC Adapter as a generic lib which works fine?

Regarding configuration for UI. UI will need configuration as any other our service. Auth is just one of piece of such configuration. It will also need to know URLs for reg-service, information about env (staging, prod, etc), etc. So, we should discuss what configuration we want to provide and how. But of course nothing should be hardcoded in JS/html.
But again it has nothing to do with KC specifics.

> i think in our case, this json is very small and it can either sit in the ui or go in the back end. personally, i prefer to have it served.. it just adds for more flexibility down the line. Perhaps one day these values will need to be extended, or perhaps one or more of these fields would be needed somewhere else.

This is a good point Tina. This is what I guessed about your motivation but wanted to confirm. The auth specific configuration would probably make sense to serve in reg-service since at least part of this config (like public keys URL) could be interesting to multiple clients (including the reg service itself). But we should not make it KC specific and we don't have to, really.

Reply:
I need to add my +1 to not having any Keycloak-specific code here, at least in the Go code.  (This is speaking from experience working on fabric8-auth)

Reply:
it's removed, going to keep this PR to just cleanup

Reply:
@alexeykazakov ok, so the Keycloak client is a generic client. But to make it completely generic, we would have to provide the client.js with our service instead of pulling it "live" from the Keycloak instance. Providing it static is not recommended by the Keycloak doc for good reasons. Pulling it live on the other side would make the process Keycloak-specific. I guess no other implementation will provide a client JS with the same API on `/auth/js/keycloak.js`. Not nitpicking here, but I think this genericness discussion is somewhat over the top.